### DISCUSSION TITLE
Efficient way to draw 1000s of feature icons

### DISCUSSION THREAD

**Discussion by tdewolff** [openlayers/openlayers#16181] [Q&A] [answered]
I've seen many similar questions here and on StackOverflow, but I haven't found any relevant information for this case or a more recent version of OpenLayers.

I need to draw 1000s of features (points) each with a fixed size vector-based image; that is it should be of a certain size irrespective of the zoom level. I've figured I need to use an Icon style image, and either using an SVG or a Canvas image (SVG seemed very slow). Performance with Canvas is OK, but not great. There is some visual stuttering on panning and zooming, not bad but still not ideal. I'm caching the canvasses, and a rough (pseudo) code of what I'm doing is below:

```js
const polygon = [...];

const cache = {};
export function getIcon(type, scale, rotation) {
  if (cache[type]) {
    return new Icon({
      img: cache[type],
      scale: scale,
      rotation: rotation,
    });
  }

  let fillColor = colors[type];
  let strokeWidth = 1.5;
  let strokeColor = [0,0,0];

  let style = new Style({
    fill: new Fill({
      color: fillColor,
    }),
    stroke: new Stroke({
      color: strokeColor,
      width: strokeWidth,
    }),
  });

  let canvas = document.createElement('canvas');
  canvas.width = 40 * window.devicePixelRatio;
  canvas.height = 52 * window.devicePixelRatio;
  let vectorContext = toContext(canvas.getContext('2d'));
  vectorContext.setStyle(style);
  vectorContext.drawGeometry(polygon);

  cache[type] = canvas;

  return new Icon({
    img: canvas,
    scale: scale,
    rotation: rotation,
  });
}

let features = [];
for (let i = 0; i < 1000; i++) {
  let center = ...;
  let scale = ...;
  let rotation = ...;
  let type = ...;
  let feature = new Feature({
    geometry: new Point(center),
  });
  feature.setStyle(new Style({
    image: getIcon(type, scale, rotation),
  }));
  features.push(feature);
}
map.getLayers().push(new VectorLayer({
  source: new Vector({
    features: features,
  }),
}));
```

**Is this the best way to draw vector-based fixed-size graphics (icons) at all positions?** Any tips are very welcome!

**Comment by ahocevar** [comment, marked as answer]
There are two things you can improve:

1. Cache the icon style, not the icon
2. Create one layer style function instead of multiple feature styles

**Reply by tdewolff** [reply]
Thank you for the advice, it runs smoother now :-)
